### ISSUE TITLE
Move cron to actual crontab

### ISSUE THREAD

**Issue by colegleason** [HabitRPG/habitica#2344]
It seems silly to have cron run on a request. It makes it hard to compute (daysSince, startOfDay), hard to debug (when did the user last log on?), and I suspect it may affect game mechanics (what happens if the user doesn't log onto habit for 3/4 days? Does the boss take damage? Do we just compute that when they log back on?).

Is there any data we need from the request that we can't have if we move cron over to an actual scheduler?

**Comment by lefnire** [comment]
definitely worth consideration. Convo started here https://github.com/HabitRPG/habitrpg/issues/2126#issuecomment-31302972

**Comment by colegleason** [comment]
Yeah, I don't think it would be too bad. And CDS/timeZone calculations aren't a big deal. I think we might actually want to add a field `user.cronTime` that translates the CDS/timezoneOffset to server time. I think that would be:

```
user.cronTime = moment().zone(timezoneOffset).set('hour', dayStart).zone(serverTZ).get('hour')
```

Then we can do `db.users.find({cronTime:moment().get('hour')})` and run the script every hour.

**Comment by lefnire** [comment]
`db.users.find({cronTime:moment().get('hour'),'stats.hp':{$gt:0}})` (to exclude inactives). I like where this is going. I suppose it's set once automatically where `timezoneOffset` is being set, and then once for every time they change CDS? 

This will also cut back on request times, since cron strikes me as pretty expensive. This strikes me as clearly the solution, I'm all in. Only con I see is testability might be tough. I won't be able to contribute to the cause till next week, so go on without me if you've got a gran scheme, but I'll dive in soon as I get back

**Comment by colegleason** [comment]
I have some ideas for logging, testing, and audits. I also think we should add an `$or` to the query to get people with `lastCron > (24 hours ago` in case a previous cron job failed for some reason. But then people might have cron run at odd times. I dunno.

**Comment by colegleason** [comment]
Actually, let's not do that. It confuses users, especially when they move their Custom Day Start time.

**Comment by snicker** [comment]
:+1:

**Comment by lefnire** [comment]
I'm actually in favor of the idea before you dismissed it, or something similar. Eg, I'm not sure how reliable Scheduler is during edge-case moments. Eg, if I do a deploy (happens maybe ~3x/day) right when cron would have hit, is it smart enough to try again? 'Cause if not, that batch of dayStart users will  get skipped for a full day. I think the `$or` solution might be the lesser of two evils in this situation

**Comment by colegleason** [comment]
Well, I think the better solution is to make the scheduler more robust. Maybe if we use a custom clock process and have it audit the users after cron to make sure it was run? I don't know.

**Comment by snicker** [comment]
snicker's thoughts on the subject:
- there should be a crontab/system scheduler task that gets executed on an hourly or half-hourly basis, as it may be expensive, depending on the tasks happening. This uses the approach that @colegleason wrote up to call from the commandline.
- the code that is executed every hour would build a collection of all users whose daystart > now and lastcron < (now - 24h),
- for each user in that collection, run all cron tasks for that particular user, and update lastcron on success.
- if any of the cron task fails, that user gets re-added to the collection with a failure count or reason. any more than X failures or if a Y failure of any kind occurs can have special action taken. (this is the build in robustness cole is talking about)

once we build this approach, we can add additional "tasks" that may need to occur on a frequency basis without having to worry about a system crontab (synchronizing between other todo apps, checking IFTTT type stuff, etc).

I always thought the current 'cron' approach had a strange codesmell, and thought that the app was trying to avoid having something in the system crontab/scheduler to avoid the complexity.This idea could still work, because you would just create a similar system to how it is now, but the cron time is tracked app-wide rather than on a per-user basis.

**Comment by colegleason** [comment]
I'm currently writing tests for this and I'm pretty confused by the test infrastructure for the node stuff. It seems the only tests that currently touch the server side code is api.mocha.coffee. The other stuff seems to be selenium tests. Is that correct, @lefnire?

**Comment by djuretic** [comment]
There is a special case, some timezones' offset is not always an integer number of hours. Example: Indian Standard Time is UTC+05:30, so that cron should be run at least every 30 minutes.

@colegleason I can answer that: 
There area three types of tests:
- Karma unit tests. No database and no webserver.
- Protractor/Selenium tests (e2e tests), testing against database and webserver via browser interaction.
- API tests against database and webserver via direct HTTP requests (api.mocha.coffee).

So go ahead and modify api.mocha.coffee.

**Comment by paglias** [comment]
The only problem I see with that, otherwise I believe it's the best solutions, is execution time. We risk a delay of some minuted between the time the user's day expires and the cron is run. Even with around 4000 active users (@lefnire ?) if we have 500 of them resetting every hour (some hours will be much crowded, others less) it will take some time to do all the calculations.

**Comment by lefnire** [comment]
@paglias [this comment](https://github.com/HabitRPG/habitrpg/pull/2346#discussion_r8953610) (just so you know I'm not ignoring your question)

**Comment by markusholdings** [comment]
I want to bring up that this might be a very bad idea.  All the issues I brought up in #1057 only apply to crontab.  In other words, If we do switch over then we might have to deal with drifting and some dailies not resetting anyway.  And if we do run into those issues, we might have to break into the crontab dependency itself and rewrite it anyway.

The way the site uses cron is not a source of issues.  The issue comes in with regards to timestamps, and the way cron is currently being used doesnt check timestamps to begin with.

**Comment by deilann** [comment]
@lefnire you said if we don't do this manual cron would be easier yeah?

**Comment by lefnire** [comment]
There are pros and cons. 

### poormanscron

(current setup, named after [drupal's similar approach](https://drupal.org/project/poormanscron)). 
- Makes manual cron much easier
- Lighter on the server, since it's calculated on demand only by current active users.
- No setup needed for local installations or installing on servers

### crontab
- I'm not sure @colegleason would agree with @markusholdings's assessment, else he might'n't have gone through all dev trouble. It's been a long while since I've seen the code, so maybe we should all sit down and codereview it to see 
- Allows for other features; eg, email notifs like "log in before tonight, you're about to die!" or "we haven't seen you for a while", etc.

**Comment by deilann** [comment]
I think in the long run, the second pro of crontab is going to be kind of
critical for Habit's success, tbh.

On the other hand, I think manual cron is also a must have for Habit's
success.

Current system makes manual cron harder, but not impossible, afaik.

**Comment by markusholdings** [comment]
@lefnire  @deilann I agree.  I am in total support of the second pro of using crontab.  I'm just worried about making the jobs too expensive, so I wouldn't recommend it be used to replace the poormanscron.

**Comment by markusholdings** [comment]
An additional suggestion, you guys have probably already figured this, but a crontab for parties would very likely fix the #2657 issue and add significant functionality to party abilities, such as the issue brought up in #2523.  And I don't think it will turn out to be terribly problematic to do so in terms of expense.

**Comment by Alys** [comment]
I'm closing this because we have plans for cron/yesterdailies that would not work if a crontab was used.
